Unprocessed Transition Cache

nholland94 opened the pull request:
GUIDE FOR REVIEW
=================

1. Read implementation of cache library in `src/lib/cache_lib`.
2. Read signature changes in `src/protocols`.
3. Read functor instantiation of transmuter cache in `src/lib/transition_handler/unprocessed_transition_cache.ml`.
4. Review all other files, which integrate calls to `Cached.transform` and `Unprocessed_transition_cache.register` where necessary.

bkase commented on src/lib/cache_lib/intf.ml:
```diff
@@ -0,0 +1,111 @@
+
+    val peek : ('t, _) t -> 't
+
+    val invalidate : ('t, _) t -> 't Or_error.t
```
can you put a doccomment here about what the `Or_error` means?

bkase commented on src/lib/cache_lib/intf.ml:
```diff
@@ -0,0 +1,111 @@
+      ('t Deferred.t, 'cache_t) t -> ('t, 'cache_t) t Deferred.t
+
+    val lift_result :
+      (('t, 'e) Result.t, 'cache_t) t -> (('t, 'cache_t) t, 'e) Result.t
```
these are typically called `sequence` (so `sequence_result` and `sequence_deferred`). see https://github.com/purescript/purescript-foldable-traversable/blob/master/src/Data/Traversable.purs#L52 (this operation is also called sequence in Haskell, Scala, Kotlin, Swift, and F# communities)

nholland94 replied:
That word feels wrong to me, but I'll trust you on this one.

bkase commented on src/lib/cache_lib/intf.ml:
```diff
@@ -0,0 +1,111 @@
+open Async_kernel
+open Core_kernel
+
+module Constant = struct
```
can you put doccomments on these modules (Constant, Cached, Cache, Transmuter, etc)

bkase commented on src/lib/cache_lib/intf.ml:
```diff
@@ -0,0 +1,111 @@
+      -> (module Hash_set.Elt_plain with type t = 'elt)
+      -> 'elt t
+
+    val register : 'elt t -> 'elt -> ('elt, 'elt) Cached.t Or_error.t
```
comment here about what the or-error means

bkase commented on src/lib/cache_lib/impl.ml:
```diff
@@ -0,0 +1,236 @@
+    let create (type elt) ~name ~logger
+        (module Elt : Hash_set.Elt_plain with type t = elt) : elt t =
+      let set =
+        Hash_set.create ~growth_allowed:true ?size:None (module Elt) ()
```
it's a bit dangerous to allow this to have an unbounded size. We should have a maximum and a policy for handling what happens when it fills up (the policy maybe should be crash for now so we can catch bugs).

bkase commented on src/lib/cache_lib/impl.ml:
```diff
@@ -0,0 +1,236 @@
+          { original: 'a
+          ; mutant: 'b
+          ; parent: ('c, 'a) t
+          ; mutable consumed: bool }
```
can you comment here about the `consumed` semantics. It wasn't obvious to me initially that you'd consume the parent whenever you transform to make a new value.

nholland94 replied:
I think my doccoments on the interface make this clearer now.

bkase commented on src/lib/cache_lib/impl.ml:
```diff
@@ -0,0 +1,236 @@
+    let transform (type a b) (t : (a, b) t) ~(f : a -> 'c) : ('c, b) t =
+      consume t ;
+      Derivative
+        {original= original t; mutant= f (value t); parent= t; consumed= false}
```
Should you also add a finalizer on the derivation? Without that, we could still be forgetting to consume derived cached values.

nholland94 replied:
good catch

bkase commented on src/lib/cache_lib/impl.ml:
```diff
@@ -0,0 +1,236 @@
+      match peek t with
+      | Ok x -> Ok (transform t ~f:(Fn.const x))
+      | Error err ->
+          (* TODO: should this log something if there was an error? *)
```
yes

bkase commented on src/lib/cache_lib/impl.ml:
```diff
@@ -0,0 +1,236 @@
+  end
+end
+
+let%test_module "cache_lib test instance" =
```
consider adding tests around transmuter/derived caches

bkase commented on src/lib/transition_frontier_controller_tests/test_ledger_catchup.ml:
```diff
@@ -43,15 +45,23 @@ let%test_module "Ledger catchup" =
           let expected_breadcrumbs =
-            Transition_frontier.path_map peer.frontier best_breadcrumb ~f:Fn.id
+            Transition_frontier.path_map peer.frontier best_breadcrumb
+              ~f:Cache_lib.Cached.phantom
```
is phantom just used to lift into the `Cached.t` context without actually caching something (for type-glueing or something)? If so, can we rename to `don't_store` or something like that (and also add some comments explaining what the phantom does)

bkase replied:
the name phantom confuses me because I expect it in the phantom type context

nholland94 replied:
well, it is basically a "value phantom" :)

I documented it in my latest commit (pushing soon), if it's still not clear there I can rename it. Don't like `don't_store` though as it changes the semantics (you don't store a `Cached.t`, a `Cached.t` is stored).

nholland94 replied:
Changed to `pure` as discussed.

bkase commented on src/lib/transition_handler/catchup_scheduler.ml:
```diff
@@ -34,47 +37,64 @@ module Make (Inputs : Inputs.S) = struct
+            match Cached.lift_result cached_breadcrumb_result with
+            | Error (`Validation_error e) ->
+                (* TODO: Punish *)
+                Logger.error logger
```
should be `Logger.faulty_peer` right?

nholland94 replied:
yes

bkase commented on src/lib/transition_handler/catchup_scheduler.ml:
```diff
@@ -101,19 +121,22 @@ module Make (Inputs : Inputs.S) = struct
     let make_timeout duration =
       Time.Timeout.create t.time_controller duration ~f:(fun _ ->
-          don't_wait_for (Writer.write t.catchup_job_writer transition) )
+          (* it's ok to create a new thread here because the thread essentially does no work *)
```
why? Does the `make_timeout` happen more infrequently than the reading from the pipe?

nholland94 replied:
I don't think I understand the question. Writing to a pipe is essentially just a block. So ignoring a block basically just means that you aren't binding after it resolves.

bkase replied:
but if this function is invoked more often than the write resolves we're queueing more async jobs

bkase commented on src/lib/transition_handler/processor.ml:
```diff
@@ -1,16 +1,27 @@
+(* This module contains the transition processor. The transition processor is
```
make this a doccomment?

bkase commented on src/lib/transition_handler/validator.ml:
```diff
@@ -58,13 +75,27 @@ module Make (Inputs : Inputs.S) :
+                    * function to only return Ok *)
+                   Unprocessed_transition_cache.register
+                     unprocessed_transition_cache transition_with_hash
+                   |> Or_error.ok_exn
```
it's kind of dirty to decouple the check and the unwrapping here -- can you instead register in the `validate_transition` and return a `_ Cached.t Or_error.t` instead of a `unit Or_error.t`?

bkase commented on src/lib/cache_lib/intf.ml:
```diff
@@ -9,29 +12,69 @@ module Constant = struct
   module type S = sig
     type ('t, 'cache_t) t
 
     val phantom : 't -> ('t, _) t
```
I think `lift` actually is a good name for this function -- you're lifting a value into the cached context, but phantom seems acceptable too

johnwu93 commented:
Overall, this looks fine. I couldn't find any serious runtime errors.

nholland94 commented:
Updated the diagram as requested. I went ahead and added a "breadcrumb builder" as a unification between the work done by catchup and catchup scheduler to build the breadcrumbs. It's not architected like that right now, but we should move towards that as there is essentially duplicated code between the catchup and catchup scheduler, plus it would allow us to put a hard limit on the number of active breadcrumb building jobs, which we don't do right now. Anyway, it made the graph easier to read.

https://github.com/CodaProtocol/coda/blob/5b160b371e649a17e707731c346e51317d1fb88c/docs/res/transition_frontier_controller.dot.png

nholland94 commented:
https://github.com/CodaProtocol/coda/issues/1718
